### PR TITLE
Sync connector 225

### PR DESCRIPTION
**Pull request by ewjoachim**
Closes #225 

Haven't touched the tests yet.

### Successful PR Checklist:

- [x] Tests
  - [ ] (not applicable?)
- [x] Documentation
  - [ ] (not applicable?)
- [x] Had a good time contributing?

### CODE REVIEW

File: procrastinate/app.py [hunk 1]
```diff
@@ -54,16 +54,22 @@ def __init__(
...
     ):
         """
         Parameters
         ----------
         connector :
-            Instance of a subclass of :py:class:`procrastinate.connector.BaseConnector`,
-            typically `AiopgConnector`. It will be responsible for all communications
+            Typically an `AiopgConnector`. It will be responsible for all communications
             with the database. Mandatory.
+        sync_defer_connector :
```
[the remark this answers is not in the document]
Comment (ewjoachim) [reply procrastinate/app.py:67]: Good catch

**Comment by ewjoachim** [comment]
Ok, I've done a 2nd pass. Documentation is stale now.

File: procrastinate/psycopg2_connector.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        }
+        final_args.update(pool_args)
+        return final_args
+
+    @wrap_exceptions
+    def close(self) -> None:
+        """
+        Close the pool
+        """
+        if self._pool:
```
Comment (ewjoachim) [inline procrastinate/psycopg2_connector.py:99]: useless if

File: procrastinate/store.py [hunk 3]
```diff
@@ -14,30 +14,54 @@ def get_channel_for_queues(queues: Optional[Iterable[str]] = None) -> Iterable[s
...
 class JobStore:
-    def __init__(self, connector: connector.BaseConnector):
+    def __init__(
+        self, connector: connector.BaseConnector,
+    ):
         self.connector = connector
 
-    async def defer_job(self, job: jobs.Job) -> int:
+    async def defer_job_async(self, job: jobs.Job) -> int:
+        # Make sure this code stays synchronized with ._defer_job_sync()
```
Comment (ewjoachim) [inline procrastinate/store.py:23]:
```suggestion
        # Make sure this code stays synchronized with .defer_job_sync()
```

File: docs/howto/sync_defer.rst [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+Pros:
+
+- Works better with multithreaded programs
+- Might be a bit faster because you don't create an event loop every time
+
+Cons:
+
+- Uses its own connection pool, which will increase the total number of connections
+  used by procrastinate.
```
Comment (ewjoachim) [inline docs/howto/sync_defer.rst:40]: In discussion

File: procrastinate/aiopg_connector.py [hunk 5]
```diff
@@ -70,6 +70,12 @@ def __init__(
             to the function used by psycopg2. See the `psycopg2 doc`_. Unused if the
             pool is externally created and set into the connector through the
             `AiopgConnector.set_pool` method.
+        real_sync_defer :
+            If a synchronous call to a defer operation is issued, whether to call a
+            really synchronous psycopg2 implementation (``True``) which will use its own
+            connection pool, or a synchronous wrapper around this asynchronous
+            connector, which may not play as nicely with multithreaded programs but will
+            use connections from this connector's pool (``False``).
```
Comment (ewjoachim) [inline procrastinate/aiopg_connector.py:78]: Link to how to or discussion

File: docs/discussions.rst [hunk 6]
```diff
@@ -149,7 +146,39 @@ and you don't want to have to maintain their equivalent using a synchronous inte
...
+When deferring a task, on the other hand, Procrastinate needs to play nice with
+your program. This is where there are choices to make.
+
+Procrastinate supports two ways of doing synchronous I/O:
+
+- "classic" synchronous I/O (using synchronous database drivers such as ``Psycopg2``).
+  This mode is necessary in multi-threaded cases.
+- "mixed" I/O (synchronously launching an event loop, and have asynchronous coroutine
+  run under the hood).
+  This mode will use less connections when used within another job.
```
Comment (elemoine) [inline docs/discussions.rst:172]: I've been thinking about this whole thing over the week-end :smile: And I must say that I am not a big fan of supporting two ways of doing synchronous I/O. The paragraph above, where we explain that there's a "classic" way and a "mixed" way, is, to me, a clear sign of complexity and poor design. I think what I'd like to see a very clear separation between the sync API and the async API. I'd even consider removing the `add_sync_api` decorator, and making the async API async only. And introducing a separate sync API for sync applications. I know my words may sound a bit theoretic… but that's all I have at this point.

Comment (ewjoachim) [reply docs/discussions.rst:172]: Ok, but what about deferring from within a job ? You already have an app and you most probably wish not to make a new pool for this.

File: docs/discussions.rst [hunk 7]
```diff
@@ -153,7 +150,37 @@ and you don't want to have to maintain their equivalent using a synchronous inte
...
+When deferring a task, on the other hand, Procrastinate needs to play nice with
+your program. This is where there are choices to make.
+
+Procrastinate supports two ways of doing synchronous I/O:
+
+- "classic" synchronous I/O (using synchronous database drivers such as ``Psycopg2``).
+  This mode is necessary in multi-threaded cases.
+- "mixed" I/O (synchronously launching an event loop, and have asynchronous coroutine
+  run under the hood). This mode is adapted for synchronously defering jobs from other
+  job, from within the workers.
```
Comment (elemoine) [inline docs/discussions.rst:176]:
```suggestion
  jobs, from within the workers.
```

File: docs/discussions.rst [hunk 8]
```diff
@@ -153,7 +150,37 @@ and you don't want to have to maintain their equivalent using a synchronous inte
...
+
+When deferring a task, on the other hand, Procrastinate needs to play nice with
+your program. This is where there are choices to make.
+
+Procrastinate supports two ways of doing synchronous I/O:
+
+- "classic" synchronous I/O (using synchronous database drivers such as ``Psycopg2``).
+  This mode is necessary in multi-threaded cases.
+- "mixed" I/O (synchronously launching an event loop, and have asynchronous coroutine
+  run under the hood). This mode is adapted for synchronously defering jobs from other
```
Comment (elemoine) [inline docs/discussions.rst:175]:
```suggestion
  run under the hood). This mode is adapted for synchronously deferring jobs from other
```

File: docs/discussions.rst [hunk 9]
```diff
@@ -265,11 +292,15 @@ this? Here are a few resources:
   with sync **and** async callers: "Just add await" from Andrew Godwin:
   https://www.youtube.com/watch?v=oMHrDy62kgE
 
+That being said, synchronous defers (see `discussion-sync-defer`) rely on a few
+specific methods that have been duplicated by hand (one async and one sync version).
```
Comment (elemoine) [inline docs/discussions.rst:296]: @ewjoachim is this comment sill valid? I am wondering about "duplicated by hand".

Comment (ewjoachim) [reply docs/discussions.rst]: I've clarified it, but it's still valid.

**Comment by ewjoachim** [comment]
This closed #152 and #149

**Pull request merged**